src/test/java/com/google/cloud/spanner/pgadapter/python/PythonTransactionTests.java

```diff
@@ -0,0 +1,1095 @@
+    statements.add(sql4);
+
+    // 1 commit request will be sent
+    statements.add("transaction");
+    statements.add("commit");
+
+    statements.add("transaction");
+    statements.add("set readonly True");
+
+    // query will be executed as expected
+    statements.add("query");
+    statements.add(sql3);
+
+    // update will throw an error because the readonly is activated
+    statements.add("update");
+    statements.add(sql2);
+
+    mockSpanner.putStatementResult(
+        StatementResult.query(Statement.of(sql1), createResultSet(1, "abcd")));
+    mockSpanner.putStatementResult(StatementResult.update(Statement.of(sql2), 1));
+    mockSpanner.putStatementResult(
+        StatementResult.query(Statement.of(sql3), createResultSet(2, "pqrs")));
+    mockSpanner.putStatementResult(StatementResult.update(Statement.of(sql4), 2));
+
+    String expectedOutput =
+        "(1, 'abcd')\n"
+            + "1\n"
+            + "2\n"
+            + "(2, 'pqrs')\n"
+            + "FAILED_PRECONDITION: Update statements are not allowed for read-only transactions\n";
+    String actualOutput = executeTransactions(pgServer.getLocalPort(), statements);
+    Assert.assertEquals(expectedOutput, actualOutput);
+
+    Assert.assertEquals(1, mockSpanner.countRequestsOfType(CommitRequest.class));
+    Assert.assertEquals(0, mockSpanner.countRequestsOfType(RollbackRequest.class));
+    Assert.assertEquals(4, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+
+    List<String> sql = new ArrayList<>();
+
+    for (ExecuteSqlRequest request : mockSpanner.getRequestsOfType(ExecuteSqlRequest.class)) {
+      sql.add(request.getSql());
+    }
+
+    Assert.assertEquals(1, Collections.frequency(sql, sql1));
+    Assert.assertEquals(1, Collections.frequency(sql, sql2));
+    Assert.assertEquals(1, Collections.frequency(sql, sql3));
+    Assert.assertEquals(1, Collections.frequency(sql, sql4));
+  }
+
+  @Test
+  public void testReadOnlySessionInTransactions() throws IOException, InterruptedException {
+    // tests readonly settings using set_session() function
+    List<String> statements = new ArrayList<>();
+    String sql1 = "Select * from some_table";
+    String sql2 = "insert into some_table(col1, col2) values(value1, value2)";
+    String sql3 = "Select * from some_table3";
+    String sql4 = "insert into some_table3(col1, col2) values(value1, value2)";
+
+    // query will be executed as expected
+    statements.add("query");
+    statements.add(sql1);
+
+    // update will be executed as expected
+    statements.add("update");
+    statements.add(sql2);
+
+    // update will be executed as expected
+    statements.add("update");
+    statements.add(sql4);
+
+    // 1 commit request will be sent
+    statements.add("transaction");
+    statements.add("commit");
+
+    statements.add("transaction");
+    statements.add("set session readonly True");
+
+    // query will be executed as expected
+    statements.add("query");
```

Comment:
Can we verify that this starts a multi-use read-only transaction. That is; the session (connection) is now in read-only mode, but not in autocommit mode. This means that the first statement after a commit should start a new read-only transaction, and all subsequent statements should use that read-only transaction. The transaction should last until the next commit/rollback. The following statement should start a new read-only transaction.

Reply:
Done